fairlearn/metrics/_metrics_engine.py

```diff
@@ -3,104 +3,124 @@
 
-    :param \*\*kwargs: Optional arguments to be passed to the `metric_function`
+    :param \*\*metric_params: Optional arguments to be passed to the `metric_function`
 
     :return: Object containing the result of applying ``metric_function`` to the entire dataset
-        and to each group identified in ``group_membership``.
-        If the ``metric_function`` returns a scalar, then additional fields are populated
-    :rtype: :class:`GroupMetricResult`
+        and to each group identified in ``sensitive_features``
+    :rtype: sklearn.utils.Bunch with the fields ``overall`` and ``by_group``
     """
     _check_array_sizes(y_true, y_pred, 'y_true', 'y_pred')
-    _check_array_sizes(y_true, group_membership, 'y_true', 'group_membership')
-    if sample_weight is not None:
-        _check_array_sizes(y_true, sample_weight, 'y_true', 'sample_weight')
-
-    result = GroupMetricResult()
+    _check_array_sizes(y_true, sensitive_features, 'y_true', 'sensitive_features')
 
     # Make everything a numpy array
     # This allows for fast slicing of the groups
-    y_a = _convert_to_ndarray_and_squeeze(y_true)
+    y_t = _convert_to_ndarray_and_squeeze(y_true)
     y_p = _convert_to_ndarray_and_squeeze(y_pred)
-    g_d = _convert_to_ndarray_and_squeeze(group_membership)
-
-    s_w = None
-    if sample_weight is not None:
-        s_w = _convert_to_ndarray_and_squeeze(sample_weight)
+    s_f = _convert_to_ndarray_and_squeeze(sensitive_features)
 
     # Evaluate the overall metric with the numpy arrays
     # This ensures consistency in how metric_function is called
-    if s_w is not None:
-        result.overall = metric_function(y_a, y_p, sample_weight=s_w, **kwargs)
-    else:
-        result.overall = metric_function(y_a, y_p, **kwargs)
+    result_overall = metric_function(
+        y_t, y_p,
+        **_check_metric_params(y_t, metric_params))
 
-    groups = np.unique(group_membership)
+    groups = np.unique(s_f)
+    result_by_group = {}
     for group in groups:
-        group_indices = (group == g_d)
-        group_actual = y_a[group_indices]
-        group_predict = y_p[group_indices]
-        group_weight = None
-        if s_w is not None:
-            group_weight = s_w[group_indices]
-            result.by_group[group] = metric_function(group_actual,
-                                                     group_predict,
-                                                     sample_weight=group_weight,
-                                                     **kwargs)
+        group_indices = (group == s_f)
+        result_by_group[group] = metric_function(
+            y_t[group_indices], y_p[group_indices],
+            **_check_metric_params(y_t, metric_params, group_indices))
+
+    return Bunch(overall=result_overall, by_group=result_by_group)
+
+
+# This loosely follows the pattern of _check_fit_params in
+# sklearn/utils/validation.py
+def _check_metric_params(y_true, metric_params, indices=None):
```

Comment:
I'd like to see some tests of this routine.

Reply:
it is being partly tested by the old `sample_weights` tests.

Reply:
okay if we hold this off for a future PR?

Reply:
Given that it's a private routine, OK. However, I do think it's doing some 'magic' so direct testing, even though it's private does make sense.

Reply:
I'll hold off. There's more magic that needs to happen in this routine to enable crossvalidation with sklearn. I'll add testing then.

Reply:
Yeah sorry about that. I'm actively working on this one on the sklearn side these days.

Reply:
@adrinjalali awesome! keep us posted.